src/torch_admp/utils.py

```
@@ -284,52 +270,73 @@ def __init__(self, units_dict: Optional[Dict] = None):
+        xx: torch.Tensor | None,
+    ) -> np.ndarray | None:
+        if xx is None:
+            return None
+        assert xx is not None
+        # Create a reverse mapping of PT_PRECISION_DICT
+        reverse_precision_dict = {v: k for k, v in PT_PRECISION_DICT.items()}
+        # Use the reverse mapping to find keys with the desired value
+        prec = reverse_precision_dict.get(xx.dtype, None)
+        prec = NP_PRECISION_DICT.get(prec, None)
+        if prec is None:
+            raise ValueError(f"unknown precision {xx.dtype}")
```

Ruff (0.14.13)

Avoid specifying long messages outside the exception class

(TRY003)